### ISSUE TITLE
NDDataArray - using weakref for uncertainty and include it in tests

### ISSUE THREAD

**Issue by MSeifert04** [astropy/astropy#4825]
In #4799 I realized that the current hardlinking between `NDData` and `uncertainty` leads to a memory leak. I don't want to update it in that PR but if the PR get's merged I (or someone) else need to edit the `uncertainty.setter` of `NDDataArray` too.

Maybe even deprecating `NDDataArray` if `NDDataRef` from #4797 is merged. But at least `ccdproc` is still using it.

Also currently there are few tests concerning `NDDataArray`. One possibility would be to include `ccdproc` and other affiliated packages that use `NDData`-derivatives into the tests or make the Mixin tests also include `NDDataArray`.

**Comment by pllim** [comment]
@MSeifert04 , do you have more info on the memory leak, say, from a diagnostic tool?

**Comment by MSeifert04** [comment]
@pllim Good point I thought the open-files issue from #4797 would be enough but I had done some tests locally, I'll submit them shortly.

**Comment by MSeifert04** [comment]
This was the quick test I originally made (slightly adapted though). The x-axis of the plots is the number of times the function is called. The y-axis gives the percentage of memory used.

For simplicity I've created a function that creates the NDDataArray but doesn't return it. That should immediatly trigger cleanup of any not returned variables. (The results also look the same if I `sleep(1)` in between.)

```
from astropy.nddata import NDDataArray, StdDevUncertainty
import numpy as np
import time
import psutil

# Function that creates NDDataArray-objects and returns the memory usage
def do_analysis():
    mem_usage = []
    for i in range(10):
        create_nddata()
        mem_usage.append(psutil.virtual_memory().percent)
    return mem_usage

mem_usage = do_analysis()

import matplotlib.pyplot as plt

plt.plot(mem_usage)
plt.show()
```

# Currently

## Without uncertainty
```
def create_nddata():
    ndd = NDDataArray(np.ones((5000,5000)))
```
![figure_1](https://cloud.githubusercontent.com/assets/14200878/14957406/88e3dd56-1085-11e6-8a60-cc347ecde3a7.png)

## With uncertainty (potential memory leak)

```
def create_nddata():
    ndd = NDDataArray(np.ones((3000,3000)),
                      uncertainty=StdDevUncertainty(np.ones((3000, 3000))))
```

![figure_1-1](https://cloud.githubusercontent.com/assets/14200878/14957457/bf581da2-1085-11e6-91c2-aa909fca28e1.png)

# Using the version from the current PR (memory leak fixed)

```
def create_nddata():
    ndd = NDDataArray(np.ones((3000,3000)),
                      uncertainty=StdDevUncertainty(np.ones((3000, 3000))))
```

![figure_1-2](https://cloud.githubusercontent.com/assets/14200878/14957519/0cc7deb0-1086-11e6-9916-ac47aa9f8291.png)

**Comment by MSeifert04** [comment]
I hope that's somehow convincing. :smile: If unclear (sorry I was in a hurry when creating these figures) I can update the plots and analysis.

**Comment by pllim** [comment]
Just want to be clear -- You mentioned two different PRs above. Which one actually has the fix?

**Comment by MSeifert04** [comment]
@pllim #4799 has the fix.

**Comment by MSeifert04** [comment]
But the PR isn't about the fix. The fix was just a way to deal with the test-failures.

**Comment by MSeifert04** [comment]
I've edited the plots a bit. Is that sufficient to conclude that there is a memory-leak?

**Comment by pllim** [comment]
Thanks for the plots! @mwcraig ?

**Comment by mwcraig** [comment]
@plim I'll get to this and the other nddata PRs on Wednesday, will try to hit this one tomorrow.

**Comment by MSeifert04** [comment]
@mwcraig If it would help you for review I could open another PR including _only_ the fix for that issue. Currently the fix is in #4799 and probably a bit "hidden".

**Comment by mwcraig** [comment]
@MSeifert04 wow, sorry it took so long to actually look at this. If you could separate the memory leak (which I think MUST get into 1.2 if at all possible) from the broader IO pR #4799 that would be AWESOME!

**Comment by MSeifert04** [comment]
@mwcraig So not splitting it would mean a higher chance of #4799 getting merged and be extra work? That doesn't sound very convincing to me. :smile: 

I'll try to create a PR only containing the memory-leak and stolen-parent fix.

**Comment by mwcraig** [comment]
Hmm, hadn't thought about it that way....

**Comment by MSeifert04** [comment]
@mwcraig PR #4799 is now without the fix (failing because of the memory leak) and PR #4862 about fixing the memory leak and associated problems.

**Comment by mwcraig** [comment]
@MSeifert04 -- is the memory leak part of this fixed in #4862? A separate issue has been opened for potential renaming/reorganization of classes.

If this is fixed now, please close (and thanks again for fixing and finding this).

**Comment by MSeifert04** [comment]
@mwcraig I'll submit a PR to make sure it's fixed for NDDataArray too.
